File: Divide_&_Conqeur/240_Search_2D_Matrix_II/solution240.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool divide_and_conquer(vector<vector<int>>& mat, int min_i, int min_j, int max_i, int max_j, int tar) {
        if (min_i > max_i || min_j > max_j) {
            return false;
        }

        int med_i = (min_i + max_i) / 2;
        int med_j = (min_j + max_j) / 2;

        if (mat[med_i][med_j] == tar) {
            return true;
        }

        if (mat[med_i][med_j] > tar) {
            if (divide_and_conquer(mat, min_i, min_j, max_i, med_j - 1, tar)) {
                return true;
            }
            if (divide_and_conquer(mat, min_i, med_j, med_i - 1, max_j, tar)) {
                return true;
            }
        } else {
            if (divide_and_conquer(mat, med_i + 1, min_j, max_i, max_j, tar)) {
                return true;
            }
            if (divide_and_conquer(mat, min_i, med_j + 1, med_i, max_j, tar)) {
                return true;
            }
        }

        return false;
    }

    bool searchMatrix(vector<vector<int>>& mat, int tar) {
        return divide_and_conquer(mat, 0, 0, mat.size() - 1, mat[0].size() - 1, tar);
    }
};

// Time Complexity:
// Space Complexity:

// Runtime1:
// Memory1:

// Runtime2:
// Memory2:

// Runtime3:
// Memory3:

File: Divide_&_Conqeur/240_Search_2D_Matrix_II/solution240.py
class Solution:
    def divide_and_conquer(self, mat, min_i, min_j, max_i, max_j, tar):
        if min_i > max_i or min_j > max_j:
            return False

        med_i = (min_i + max_i) // 2
        med_j = (min_j + max_j) // 2
        if mat[med_i][med_j] == tar:
            return True
        if mat[med_i][med_j] > tar:
            if self.divide_and_conquer(mat, min_i, min_j, max_i, med_j - 1, tar):
                return True
            if self.divide_and_conquer(mat, min_i, med_j, med_i - 1, max_j, tar):
                return True
        else:
            if self.divide_and_conquer(mat, med_i + 1, min_j, max_i, max_j, tar): 
                return True
            if self.divide_and_conquer(mat, min_i, med_j + 1, med_i, max_j, tar): 
                return True

        return False
        
    def searchMatrix(self, mat: List[List[int]], tar: int) -> bool:
        return self.divide_and_conquer(mat, 0, 0, len(mat)-1, len(mat[0])-1, tar)

# Runtime1: 140ms, beats 54.08%
# Memory1: 23MB, beats 45.08%

# Runtime2: 134ms, beats 83.27%
# Memory2: 22.92MB, beats 45.08% 

# Runtime3: 121ms, beats 99.48%
# Memory3: 22.96MB, beast 45.08%